File: Algoexpert/Shift Linked List/python/shift_linked_list.py
import unittest

class LinkedList:
    def __init__(self, value):
        self.value = value
        self.next = None

def linkedListToArray(head):
    array = []
    current = head
    while current is not None:
        array.append(current.value)
        current = current.next
    return array

def shiftLinkedList(head, k):
	
	listTail = head
	length = 1
	while listTail.next:
		listTail = listTail.next
		length += 1
	
	offset = abs(k) % length
	if offset == 0:
		return head
	
	newTailPosition = length - offset if k > 0 else offset
	
	newTail = head
	for _ in range(1, newTailPosition):
		newTail = newTail.next
	
	newHead = newTail.next
	newTail.next = None
	listTail.next = head
	
	return newHead

class TestProgram(unittest.TestCase):
    def test_case_1(self):
        head = LinkedList(0)
        head.next = LinkedList(1)
        head.next.next = LinkedList(2)
        head.next.next.next = LinkedList(3)
        head.next.next.next.next = LinkedList(4)
        head.next.next.next.next.next = LinkedList(5)
        result = shiftLinkedList(head, 2)
        array = linkedListToArray(result)

        expected = [4, 5, 0, 1, 2, 3]
        self.assertEqual(expected, array)

if __name__ == "__main__":
    unittest.main()

File: Algoexpert/Shift Linked List/cpp/shift_linked_list.cpp
#include <bits/stdc++.h>

using namespace std;

class LinkedList {
    public:
        int value;
        LinkedList *next;

        LinkedList(int value) {
            this->value = value;
            next = nullptr;
        }
};

vector<int> linkedListToArray(LinkedList *head) {
    vector<int> array{};
    auto current = head;
    while (current != nullptr) {
        array.push_back(current->value);
        current = current->next;
    }
    return array;
}

LinkedList *shiftLinkedList(LinkedList *head, int k) {

    int listLength = 1;
    LinkedList *listTail = head;
    while (listTail->next != nullptr) {
        listTail = listTail->next;
        listLength++;
    }

    int offset = abs(k) % listLength;
    if (offset == 0) return head;
    int newTailPosition = k > 0 ? listLength - offset : offset;
    LinkedList *newTail = head;
    for (int i = 1; i < newTailPosition; i++) {
        newTail = newTail->next;
    }
    LinkedList *newHead = newTail->next;
    newTail->next = nullptr;
    listTail->next = head;
    return newHead;
}

int main() {

    auto head = new LinkedList(0);
    head->next = new LinkedList(1);
    head->next->next = new LinkedList(2);
    head->next->next->next = new LinkedList(3);
    head->next->next->next->next = new LinkedList(4);
    head->next->next->next->next->next = new LinkedList(5);
    auto result = shiftLinkedList(head, 2);
    auto array = linkedListToArray(result);

    vector<int> expected{4, 5, 0, 1, 2, 3};
    assert(expected == array);

}
